x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -155,4 +194,128 @@ private void ensureEnabled() {
+     * indices privileges from the base descriptors that match and creates a
+     * {@code bool} query with a {@code should} clause and
+     * {@code minimum_should_match} set to 1. For the indices privileges from
+     * the child role descriptors that match, we add them as {@code should}
+     * clause with {@code minimum_should_match} set to 1. The base {@code bool}
+     * query is set as {@code filter} clause on this outer {@code bool} query.
+     *
+     * @param childDescriptors list of {@link RoleDescriptor} from api key
+     * request
+     * @param baseDescriptors list of {@link RoleDescriptor} for current logged
+     * in user
+     * @param result {@link SubsetResult} as determined by the
+     * {@link Role#isSubsetOf(Role)} check
+     * @param user {@link User} current logged in user
+     * @return returns list of role descriptors equal to child descriptors if
+     * they are not modified else returns modified list of role descriptors
+     * @throws IOException thrown in case of invalid query template or if unable
+     * to parse the query
+     */
+    private List<RoleDescriptor> modifyRoleDescriptorsToMakeItASubset(final List<RoleDescriptor> childDescriptors,
+            final List<RoleDescriptor> baseDescriptors, final SubsetResult result, final User user) throws IOException {
+        final Map<Set<String>, BoolQueryBuilder> indexNamePatternsToBoolQueryBuilder = new HashMap<>();
+        for (Set<String> indexNamePattern : result.setOfIndexNamesForCombiningDLSQueries()) {
+            final Automaton indexNamesAutomaton = Automatons.patterns(indexNamePattern);
+            final BoolQueryBuilder parentFilterQueryBuilder = QueryBuilders.boolQuery();
+            // Now find the index name patterns from all base descriptors that
+            // match and combine queries
+            for (RoleDescriptor rdbase : baseDescriptors) {
+                for (IndicesPrivileges indicesPriv : rdbase.getIndicesPrivileges()) {
+                    if (Operations.subsetOf(indexNamesAutomaton, Automatons.patterns(indicesPriv.getIndices()))) {
+                        final String templateResult = SecurityIndexSearcherWrapper.evaluateTemplate(indicesPriv.getQuery().utf8ToString(),
+                                scriptService, user);
+                        try (XContentParser parser = XContentType.JSON.xContent().createParser(xContentRegistry,
+                                LoggingDeprecationHandler.INSTANCE, templateResult)) {
+                            parentFilterQueryBuilder.should(AbstractQueryBuilder.parseInnerQueryBuilder(parser));
+                        }
+                    }
+                }
+            }
+            parentFilterQueryBuilder.minimumShouldMatch(1);
+
+            final BoolQueryBuilder outerBoolQueryBuilder = QueryBuilders.boolQuery();
+            outerBoolQueryBuilder.filter(parentFilterQueryBuilder);
+            // Iterate on child role descriptors and combine queries if the
```

Comment:
maybe instead of child, say descriptors that should be a subset of existing descriptors